glusterd: Removing redundant NULL checks for this and other cleanups.

schaffung opened the pull request:
Issue: It has been noticed that the NULL checks performed
on `this` are actually being done on the result of `__glusterfs_this_location()`.

Code Change:
1. Basically removing redundant the validations and assertion
functions which check if `this` is NULL.
2. Now that we have started cleaning up, the patch also tries to clean up the approach of using
`this` and `THIS` so as to minimize the variable creation (i.e. `this` ) as well as function call ( i.e. `THIS` ).


Updates: #1596
Signed-off-by: srijan-sivakumar <ssivakum@redhat.com>

Commit:
glusterd: Removing redundant NULL checks for this

Issue: It has been noticed that the NULL checks performed
on `this` are actually being done on `THIS` as `this` is
derived from `THIS`. If the `THIS` had been NULL, the
crash would have happened before itself.

Fix: Basically removing the validations and assertion
functions which check if `this` is NULL.

Fixes: #1596
Signed-off-by: srijan-sivakumar <ssivakum@redhat.com>

nik-redhat commented:
If I am not wrong this change is only for the glusterd component, right?
Don't we want to fix this in other components too?

schaffung commented:
> If I am not wrong this change is only for the glusterd component, right?
> Don't we want to fix this in other components too?

IIUC, you meant to ask if there will be commits for other xlators too?

schaffung commented:
The idea is to do it one by one. This is focused on glusterd xlator, we have other xlators too wherein we have the same issue. That'll be part of a separate patch..didn't want to get lost with this and This for this one patch itself :p

nik-redhat commented:
Ack.
Just wanted to confirm that part. So, I think it's better to update the commit message to `Updates:` instead of `Fixes:`, what say?

schaffung commented:
/recheck smoke

mykaul commented:
How can THIS return NULL? It is calling __glusterfs_this_location() which eventually will return the address of
```
xlator_t global_xlator;
```

Which is defined globally?

mykaul commented:
The ASSERT is also a debug build only statement, so pretty harmless.
All those 'VALIDATE_OR_GOTO(this, err);' across the FOPS are more wasteful, IMHO. I think they are unnecessary.

schaffung commented:
> How can THIS return NULL? It is calling __glusterfs_this_location() which eventually will return the address of
> 
> ```
> xlator_t global_xlator;
> ```
> 
> Which is defined globally?

Ack. Will change the commit statement to reflect that..

Sheetalpamecha commented on xlators/mgmt/glusterd/src/glusterd-bitd-svc.c:
```diff
@@ -149,9 +148,6 @@ glusterd_bitdsvc_reconfigure()
     glusterd_conf_t *priv = NULL;
     gf_boolean_t identical = _gf_false;
 
-    this = THIS;
```
Why we are removing this line?

schaffung replied:
Ack..it shouldn't be removed. Will add it in the next commit..

Sheetalpamecha commented on xlators/mgmt/glusterd/src/glusterd-mountbroker.c:
```diff
@@ -260,10 +260,11 @@ make_georep_mountspec(gf_mount_spec_t *mspec, const char *volnames, char *user,
     int vc = 0;
     int i = 0;
     int ret = 0;
+    xlator_t *this = THIS;
```
I don't thing we need again new var and assignment to THIS. why cant we use THIS in gf_smsg?

schaffung replied:
Technically, yes we are creating one more variable for this. The idea was to not to have THIS inside function calls.

schaffung replied:
Seeing the overall consensus, I think your suggestion was prudent. Will go with THIS in here..

mykaul replied:
Yes, I believe @Sheetalpamecha is right - it's mostly because we only call it once, and we only call it once in error paths - so I think it's the right direction.

Commit:
Made changes wrt review comments received.

Fixes: #1596
Signed-off-by: srijan-sivakumar <ssivakumar@redhat.com>

schaffung commented:
/recheck smoke

schaffung commented:
@gluster/gluster-maintainers could you please run the regression on this PR. Thank you in advance.

itisravi commented:
/run regression

mykaul commented on xlators/mgmt/glusterd/src/glusterd-brick-ops.c:
```diff
@@ -1689,7 +1682,6 @@ glusterd_remove_brick_validate_bricks(gf1_op_commands cmd, int32_t brick_count,
     glusterd_conf_t *priv = THIS->private;
```
Hmm... perhaps this should be initialized later, after line 1684?

schaffung replied:
Ack.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-gfproxyd-svc.c:
```diff
@@ -141,7 +140,6 @@ glusterd_gfproxydsvc_create_volfile(glusterd_volinfo_t *volinfo)
     xlator_t *this = NULL;
 
     this = THIS;
-    GF_VALIDATE_OR_GOTO("glusterd", this, out);
```
nit: please fix the 'if' in line 152 ('this ? this->name' part in the gf_msg_debug() call)  (this is happening in other functions as well)

nit2: we don't need the 'out' label anymore.

schaffung replied:
Ack.  Will make the changes accordingly..

mykaul commented on xlators/mgmt/glusterd/src/glusterd-svc-helper.c:
```diff
@@ -360,7 +353,7 @@ glusterd_volume_svc_check_topology_identical(
 
-    if ((!identical) || (!this) || (!this->private)) {
+    if ((!identical) || (!this->private)) {
         gf_smsg(THIS->name, GF_LOG_ERROR, errno, GD_MSG_INVALID_ARGUMENT, NULL);
```
nit: this->name

schaffung replied:
Ack..seeing too many this and THIS is not good for eyes I guess :p

mykaul commented on xlators/mgmt/glusterd/src/glusterd-syncop.c:
```diff
@@ -224,7 +224,6 @@ glusterd_syncop_aggr_rsp_dict(glusterd_op_t op, dict_t *aggr, dict_t *rsp)
     xlator_t *this = NULL;
 
     this = THIS;
```
So 'this' is used exactly once in this function - in an error log. Do we really need it at all?
Instead of:
```
gf_msg(this->name, GF_LOG_ERROR, 0,
                       GD_MSG_BRICK_MOUNDIRS_AGGR_FAIL,
                       "Failed to "
                       "aggregate brick mount dirs");
```
Let's do:
```
gf_msg(THIS->name, GF_LOG_ERROR, 0,
                       GD_MSG_BRICK_MOUNDIRS_AGGR_FAIL,
                       "Failed to "
                       "aggregate brick mount dirs");
```

And get rid of the xlator_t* parameter and the assignment.

schaffung replied:
Ack..I'll go through the management layer one more time to see if there are such instances which I've missed.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-tierd-svc-helper.c:
```diff
@@ -107,7 +107,6 @@ glusterd_svc_check_tier_volfile_identical(char *svc_name,
 
     this = THIS;
 
-    GF_VALIDATE_OR_GOTO(THIS->name, this, out);
```
The whole function - glusterd_svc_check_tier_volfile_identical(), seems to be dead code...

schaffung replied:
I think I missed this...will try to address this in the a different PR.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-volgen.c:
```diff
@@ -91,7 +91,6 @@ xlator_instantiate_va(const char *type, const char *format, va_list arg)
     char *volname = NULL;
     int ret = 0;
     xlator_t *this = THIS;
```
nit: 'this' is ONLY used in error log messages. So in the positive flow, there's no point of even calling 'THIS'. I'd argue that functions such as this, should not be bothering with this unconditional assignment.

(However, let's do it in a different patch...)

mykaul commented on xlators/mgmt/glusterd/src/glusterd.c:
```diff
@@ -451,7 +448,6 @@ glusterd_rpcsvc_options_build(dict_t *options)
     int ret = 0;
     uint32_t backlog = 0;
     xlator_t *this = THIS;
-    GF_ASSERT(this);
```
nit: 'this' is only used once in this function, in an error log message. Just delete this variable and call the gf_msg() with 'THIS->name' instead.

schaffung replied:
Ack.

mykaul reviewed (commented):
Good cleanup, and opens the door to continued cleanup of needlessly calling 'THIS'. This is very evident in functions that only use 'this->name' in error flows.

schaffung commented:
/recheck smoke

Commit:
glusterd: The efficient usage of `THIS` and `this`.

This commit addresses the review comments and tries to
change code in more places wherein the `THIS` and `this` can
be handled efficiently.

Signed-off-by: srijan-sivakumar <ssivakumar@redhat.com>

Commit:
Updated commit to address review comments.

Updates: #1596
Signed-off-by: srijan-sivakumar <ssivakumar@redhat.com>

schaffung commented:
@gluster/gluster-maintainers could you please run the regression on this PR. Thank you in advance.

Commit:
Addressing Review comments.

Updates: #1596
Signed-off-by: srijan-sivakumar <ssivakumar@redhat.com>

mykaul commented on xlators/mgmt/glusterd/src/glusterd-geo-rep.c:
```diff
@@ -79,12 +79,10 @@ static char *gsync_no_restart_opts[] = {"checkpoint", "log_rsync_performance",
     int ret;
 
-    this = THIS;
-    ret = dict_get_str(this->options, "transport.address-family", &af);
```
Half of the functions in this file are static, that call one each other. We could have reduced the number of calls to THIS by passing the result between functions. We can save it for a future iteration of course.

schaffung replied:
Ack. I'll add this in the issue so that it can be tracked..

mykaul commented on xlators/mgmt/glusterd/src/glusterd-geo-rep.c:
```diff
@@ -699,10 +683,7 @@ glusterd_query_extutil_generic(char *resbuf, size_t blen, runner_t *runner,
-
-    this = THIS;
-    GF_ASSERT(this);
+    xlator_t *this = THIS;
```
nit: 'this' is only used in this function in error messages, and then they bail out. You could have replaced each 'this->name' with 'THIS->name' in this function and avoid an unconditional call to 'THIS' here.

For a follow up patch series, not in this one of course...

schaffung replied:
Ack. Adding it in the issue as comment.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-geo-rep.c:
```diff
@@ -1111,10 +1086,7 @@ glusterd_gsync_volinfo_dict_set(glusterd_volinfo_t *volinfo, char *key,
-
-    this = THIS;
-    GF_ASSERT(this);
+    xlator_t *this = THIS;
```
Same comment as above + it could have been passed from the single caller of this function - glusterd_set_gsync_knob() - which already does 'this = THIS'  (which by itself is called only from glusterd_set_gsync_confs() - which is called only from glusterd_op_gsync_create() which also does 'this = THIS;'

In short, lots of opportunities to clean this up. Not for this series of patches of course.

schaffung replied:
Ack. Same as above..

mykaul commented on xlators/mgmt/glusterd/src/glusterd-handshake.c:
```diff
@@ -189,7 +187,6 @@ glusterd_get_client_per_brick_volfile(glusterd_volinfo_t *volinfo,
     int32_t ret = -1;
 
-    GF_VALIDATE_OR_GOTO("glusterd", THIS, out);
     priv = THIS->private;
```
nit: OK, glusterd_get_client_per_brick_volfile() should be a static function. It is only called from build_volfile_path() - and we don't need 'this' here, we only need priv, which we can pass to, right?
You can then remove the validation in the following line, as we know priv is not null already.

I wouldn't do it in this patch series which is already very big...

schaffung replied:
I think the conversion of functions to static will itself be a separate patch as I've seen multiple instances of function not being static even if they should've been. 

And yes, there are places wherein only `private` is being used instead of any other component of `this`. I've added this point to the issue for further cleanup.

mykaul commented on xlators/mgmt/glusterd/src/glusterd-handshake.c:
```diff
@@ -1296,15 +1282,12 @@ gd_validate_mgmt_hndsk_req(rpcsvc_request_t *req, dict_t *dict)
     };
     glusterd_peerinfo_t *peer = NULL;
-    xlator_t *this = NULL;
+    xlator_t *this = THIS;
```
nit: just pass this from the single caller of this function - __glusterd_mgmt_hndsk_versions()  (remember to make this function static)

schaffung replied:
Ack.

mykaul commented:
I've tried to review the patches. Similar to what I used to do in my patches (when I used to actively contribute to Gluster), they are very big and cover too many files. This has annoyed the community a bit ;-)

Moreover, as I've shown in several comments, they are somewhat incomplete, in either one or two or more ways:
1. The functions only use 'this' in error paths and then bail - so just use 'THIS->name' in those gf_msg/gf_smsg/gf_log/... and don't bother with the assignment (and therefore the variable definition) at all.
2. Those functions could have easily received 'this' from their caller as they could have been defined static (many, many places!)
3. They never really needed 'this' in the first place. They needed 'priv' (this is ALL around the dht code!). Combine it with #1, and you'll see how much you can actually save.

mohit84 commented:
/run regression

Sheetalpamecha commented:
/run regression

Commit:
Made changes after regression failure.

Updates: #1596
Signed-off-by: srijan-sivakumar <ssivakumar@redhat.com>

schaffung commented:
/run regression

nik-redhat reviewed (dismissed):
I think we are good here, and the further enhancements can be addressed in a separate PR as it has already become a huge PR (~4k lines of change)!

schaffung commented:
/run regression

schaffung commented:
/run regression

schaffung commented on xlators/mgmt/glusterd/src/glusterd-rpc-ops.c:
```diff
@@ -1517,8 +1500,8 @@ glusterd_rpc_friend_add(call_frame_t *frame, xlator_t *this, void *data)
 
-    if (!frame || !this || !data) {
-        gf_smsg(THIS->name, GF_LOG_ERROR, errno, GD_MSG_INVALID_ARGUMENT, NULL);
+    if (!frame | !data) {
```
:/

Commit:
One has to be careful while working with c

Instead of a `||` operation, the cleanup left out
with `|`. Does the compiler ceck for these things?

Updates: #1596
Signed-off-by: srijan-sivakumar <ssivakumar@redhat.com>

schaffung commented:
/run regression

schaffung commented:
> I think we are good here, and the further enhancements can be addressed in a separate PR as it has already become a huge PR (~4k lines of change)!

Yeah...Having a huge code change makes it difficult to find out what causes failure in the regressions.. Like a needle in haystack.

nik-redhat reviewed (dismissed):
+1

Sheetalpamecha commented on xlators/mgmt/glusterd/src/glusterd-quota.c:
```diff
@@ -583,7 +568,7 @@ glusterd_quota_get_default_soft_limit(glusterd_volinfo_t *volinfo,
+        gf_msg(THIS->name, GF_LOG_ERROR, 0, GD_MSG_DICT_SET_FAILED,
                "Failed to set default "
                "soft-limit into dict");
         goto out;
```
nit. We can remove the label out and ret = 0 assignment

schaffung replied:
This is actually a common issue across the codebase. I haven't taken up this activity in this patch as it'd just go on to increase the side of the patch. Will address in subsequent patches.

Sheetalpamecha replied:
yes, I know. Lets take such changes in next patch.

Sheetalpamecha replied:
Maybe next time we can take file by file or few files per xlator. Else it gets too big for reviewing.

schaffung replied:
Yes, actually we haven't done the `this` and `THIS` cleanup for all xlators and as Yaniv had mentioned, there's still an opportunity to optimize things here..

schaffung commented:
/run brick-mux regression

schaffung commented:
/run brick-mux regression

schaffung commented:
/run regression

schaffung commented:
/run brick-mux regression

schaffung commented:
/run brick-mux regression

schaffung commented:
/run regression

schaffung commented:
/run brick-mux regression

schaffung commented:
/recheck smoke

schaffung commented:
/run brick-mux regression

rkothiya reviewed (commented):
+1 LGTM

schaffung commented:
/run regression

schaffung commented:
/recheck smoke

nik-redhat reviewed (commented):
@srijan-sivakumar there are few conflicts with the PR.

schaffung commented:
> @srijan-sivakumar there are few conflicts with the PR.

Ack. Will look into it.

Commit:
Merge branch 'devel' into issue1596

schaffung commented:
/run regression

Commit:
Fixing clang-format issues.

Change-Id: I68c52249af66080f59f57e558901f2654bd43cd8
Updates: #1596
Signed-off-by: srijan-sivakumar <ssivakumar@redhat.com>

schaffung commented:
/run regression

schaffung commented:
/run regression

schaffung commented:
/run regression

mohit84 commented on xlators/mgmt/glusterd/src/glusterd-brick-ops.c:
```diff
@@ -2130,17 +2107,10 @@ glusterd_op_add_brick(dict_t *dict, char **op_errstr)
-    GF_ASSERT(this);
-
-    priv = this->private;
-    GF_ASSERT(priv);
```
We should not remove GF_ASSERT check for priv, in case of THIS is NULL glusterfs_this_location sets global xlator to THIS and global_xlator does not have valid private object.

mohit84 commented:
/run brick-mux regression
